### PR TITLE
Jetpack Connect: abstract an HoC that handles connection redirects for JP sites.

### PR DESCRIPTION
**Pull request by AnnaMag**
#### Changes proposed in this Pull Request
* move the Jetpack connection functionality to an HoC avoid code duplication.

This is a preparation for the WP.com/JP [Jetpack Search purchase integration
](https://github.com/Automattic/wp-calypso/pull/41748) to handle non-Jetpack sites and move away from the `jetpack/connect` route. Purchasing a product for the Jetpack sites includes connection checks, shared with the Jetpack Connect Flow.  

The change is substantial, so for ease of review, this PR avoids major re-writes e.g. introducing Redux. Any further improvements can be handled separately. 

#### Testing instructions

Connection Flow should change for none of the Jetpack statuses.

* For any Jetpack plan, start the connection flow, e.g. `/jetpack/connect/premium` and verify that form copy, footer, input box, redirects are correct and functional, and that the flow leads to checkout.  In order to verify that the plugin was correctly setup, to be tested with a Jetpack site that is unconnected, plugin deleted, or inactive.
* the flow starting at `/jetpack/connect` without plan specified should end at the canonical plans page.

### CODE REVIEW

**Comment by gibrown** [comment]
Tested this a bit and everything I was (kinda randomly) testing seemed to work fine. Should we get a code review from someone in Jetpack on this refactor though?

**Comment by AnnaMag** [comment]
unrelated test failures after most recent rebase.

File: client/jetpack-connect/main.jsx [hunk 1]
```diff
@@ -388,17 +143,15 @@ export class JetpackConnectMain extends Component {
...
 
 		return (
 			<MainWrapper>
 				{ this.renderLocaleSuggestions() }
 				<div className="jetpack-connect__site-url-entry-container">
 					<MainHeader type={ type } />
-
 					{ this.renderSiteInput( status ) }
-					{ this.renderFooter() }
+					{ renderFooter() };
```
Comment (mattgawarecki) [inline client/jetpack-connect/main.jsx:154]: This is the nitpick-iest of nitpicks, but did you mean to have the trailing `;` here?

**Review by mattgawarecki — APPROVED** [review]
I ran through the connect flow several times. All the copy looks as I would expect, and I didn't notice anything broken. Going through flows for a specific plan ended up at checkout, and without a plan ended up at the Plans page.

I thoroughly reviewed the code as well, and other than one possible stray semi-colon in a `render` function, I think it's good to go. 👍

**Pull request merged**